adapter/internal/operator/controllers/cp/application_controller.go

```diff
@@ -0,0 +1,187 @@
+/*
+ *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package cp
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/wso2/apk/adapter/internal/discovery/xds"
+	"github.com/wso2/apk/adapter/internal/loggers"
+	"github.com/wso2/apk/adapter/pkg/logging"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cpv1alpha1 "github.com/wso2/apk/adapter/internal/operator/apis/cp/v1alpha1"
+	"github.com/wso2/apk/adapter/pkg/discovery/api/wso2/discovery/subscription"
+)
+
+// ApplicationReconciler reconciles a Application object
+type ApplicationReconciler struct {
+	client           client.Client
+	Scheme           *runtime.Scheme
+	applicationCache applicationCache
+}
+
+type applicationCache map[types.NamespacedName]*cpv1alpha1.Application
+
+// NewApplicationController creates a new Application controller instance.
+func NewApplicationController(mgr manager.Manager) error {
+	r := &ApplicationReconciler{
+		client:           mgr.GetClient(),
+		applicationCache: applicationCache{},
+	}
+	c, err := controller.New("Application", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprint("Error creating Application controller. ", err.Error()),
+			Severity:  logging.BLOCKER,
+			ErrorCode: 5001,
+		})
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &cpv1alpha1.Application{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		loggers.LoggerAPKOperator.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprint("Error watching Application resources. ", err.Error()),
+			Severity:  logging.BLOCKER,
```

Comment:
also update the sheet https://docs.google.com/spreadsheets/d/1VH2rMZ-PdPbFw6VjUWV0o1Y0gxKANuVhgUpdYD1Iiog/edit#gid=189800195 with new codes